### PR TITLE
Refactor riscv plugin from RAnal to RArch

### PR DESCRIPTION
**Pull request by as0ler**
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

### CODE REVIEW

File: libr/arch/p/arch_riscv.c [hunk 1]
```diff
@@ -2,13 +2,14 @@
 
 #include <r_lib.h>
 #include <r_asm.h>
-#include <r_anal.h>
+#include <r_arch.h>
 #include "../../asm/arch/riscv/riscv-opc.c"
```
Comment (trufae) [inline libr/arch/p/arch_riscv.c:6]: those includes should be. moved into libr/arch/p/riscv/*.c

File: libr/arch/p/arch_riscv.c [hunk 2]
```diff
@@ -2,13 +2,14 @@
...
 #include "../../asm/arch/riscv/riscv.c"
```
Comment (trufae) [inline libr/arch/p/arch_riscv.c:7]: its never a good idea to include a c file, as long as those should be dynamically linked as separate objects. but its fine for now. we will do more reviews

File: libr/arch/p/arch_riscv.c [hunk 3]
```diff
@@ -341,29 +339,31 @@ static char *riscv_disassemble(RAnal *a, ut64 addr, const ut8 *buf, int len) {//
...
+static int riscv_decode(RArchSession *s, RAnalOp *op, RArchDecodeMask mask) {
+	ut64 addr = op->addr;
+	const ut8 *buf = op->bytes;
+	int len = op->size;
 	const int no_alias = 1;
 	riscv_args_t args = {0};
 	ut64 word = 0;
-	int xlen = anal->config->bits;
-	op->addr = addr;
+	int xlen = s->config->bits;
```
Comment (trufae) [inline libr/arch/p/arch_riscv.c:349]:
```suggestion
	const int xlen = s->config->bits;
```

File: libr/arch/p/arch_riscv.c [hunk 4]
```diff
@@ -945,23 +949,23 @@ static int archinfo(RAnal *anal, int q) {
...
 	.desc = "RISC-V analysis plugin",
 	.license = "GPL",
 	.arch = "riscv",
 	.endian = R_SYS_ENDIAN_LITTLE | R_SYS_ENDIAN_BIG,
-	.bits = 32 | 64,
-	.op = &riscv_op,
-	.archinfo = archinfo,
-	.opasm = &riscv_opasm,
-	.get_reg_profile = &get_reg_profile,
+	.bits = R_SYS_BITS_PACK2(32, 64),
```
Comment (trufae) [inline libr/arch/p/arch_riscv.c:958]:
```suggestion
	.bits = R_SYS_BITS_PACK2 (32, 64),
```

File: libr/arch/p/arch_riscv.c [hunk 5]
```diff
@@ -75,7 +73,7 @@ static void get_riscv_args(riscv_args_t *args, const char *d, insn_t l, ut64 pc)
 			d++;
 			switch (*d) {
 			case 'd':
-				snprintf (RISCVARGN (args), RISCVARGSIZE , "%d", rd);
+				RISCVPRINTF ("%d", rd);
```
Comment (trufae) [inline libr/arch/p/arch_riscv.c:76]: replace this in all the other snprintf calls

**Review by trufae — CHANGES_REQUESTED** [review]
cant find the reason of the failing tests without running the code. so feel free to fix the comments. and i would go by running r2 on any failing test to find the root reason of the failure later

File: libr/arch/p/arch_riscv.c (hunk 1 above)
Comment (as0ler) [reply libr/arch/p/arch_riscv.c:6]: Fix here: 2158f3cd11d830eea37c410a4840a942e5d0e086

File: libr/arch/p/arch_riscv.c (hunk 5 above)
Comment (as0ler) [reply libr/arch/p/arch_riscv.c:76]: Fixed!

**Comment by trufae** [comment]
The failing tests are because of a bug @swoops have found in rarch! nice combo! so it should be good to merge now

**Comment by trufae** [comment]
superseeded by https://github.com/radareorg/radare2/pull/21084

**Comment by trufae** [comment]
So swifty! good contrib

**Pull request closed without merging**